staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -264,19 +450,216 @@ func (q *delayingType) waitingLoop() {
 	}
 }
 
-// insert adds the entry to the priority queue, or updates the readyAt if it already exists in the queue
-func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
-	// if the entry already exists, update the time only if it would cause the item to be queued sooner
+// insert adds the entry to the 'waiting' queue, or updates the readyAt if it already exists in the queue
+// It returns a boolean to indicate that the entry was inserted or not.
+func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) bool {
+	// if the entry already exists, update the time according to the entry.options.WhenWaiting
 	existing, exists := knownEntries[entry.data]
 	if exists {
-		if existing.readyAt.After(entry.readyAt) {
-			existing.readyAt = entry.readyAt
-			heap.Fix(q, existing.index)
+		switch entry.options.WhenWaiting {
+		case TakeShorter:
+			if existing.readyAt.After(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
+		case TakeLonger:
+			if existing.readyAt.Before(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
+		case TakeIncoming:
+			if !existing.readyAt.Equal(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
 		}
-
-		return
+		return false
 	}
 
 	heap.Push(q, entry)
 	knownEntries[entry.data] = entry
+	return true
+}
+
+// calculateNextReadyAt recalculates the nextReadyAtTime as follows:
+// - removes the timer when the 'waiting' queue is empty.
+// - resets the timer when the head 'waiting' item has changed.
+// - creates a new timer when 1+ items are queued to an empty 'waiting' queue.
+func (q *delayingType) calculateNextReadyAt(w *waitForPriorityQueue) {
+	if w.Len() == 0 {
+		// remove timer when the queue becomes empty
+		if q.nextReadyHead != nil {
+			q.nextReadyAtTimerCh = make(<-chan time.Time)
+			if q.nextReadyAtTimer != nil {
+				q.nextReadyAtTimer.Stop()
+				q.nextReadyAtTimer = nil
+			}
+			q.nextReadyHead = nil
+		}
+		return
+	}
+
+	// Keep the existing nextReadyAt timer if the head duration has not changed.
+	entry := w.Peek().(*waitFor)
+	if q.nextReadyHead != nil && q.nextReadyHead.readyAt == entry.readyAt {
+		return
+	}
+
+	// Recalculate the nextReadyAtTimer
+	q.nextReadyAtTimerCh = make(<-chan time.Time)
+	if q.nextReadyAtTimer != nil {
+		q.nextReadyAtTimer.Stop()
+		q.nextReadyAtTimer = nil
+	}
+
+	now := q.clock.Now()
+	q.nextReadyAtTimer = q.clock.NewTimer(entry.readyAt.Sub(now))
+	q.nextReadyAtTimerCh = q.nextReadyAtTimer.C()
+	q.nextReadyHead = entry
```

Comment:
can you please add a test that this behaves correctly?

Reply:
I agree that it is better to concentrate on the readAt time.Time rather than a pointer to the entry so I have refactored.  I'm not sure of your point regarding changing on the heap as the entry is always passed around as a pointer.

Reply:
> I'm not sure of your point regarding changing on the heap as the entry is always passed around as a pointer.

yes, but if you would change the readyAt of the entry in the head to an earlier time (via TakeShorter) it would change the  readyAt time in the nextReadyHead. In that case it would skip the recomputation of the timer in the early escape and the original timer would fire late.

The current solution that passes the readyAt by value, should work fine without these issues. Thanks for adding the test. Can you add the same item to the queue with TakeShorter so we can test the update of the head entry?